src/relayFeeCalculator/relayFeeCalculator.ts

```diff
@@ -0,0 +1,118 @@
+    this.coingecko = new Coingecko();
+  }
+  /**
+   * getGasCosts. This should calculate how much of the native gas token is used for a relay on the destination chain.
+   *
+   * @returns {Promise<BigNumberish>}
+   */
+  async getGasCosts(): Promise<BigNumberish> {
+    const { gasPrice, maxFeePerGas } = await this.provider.getFeeData();
+    const price = maxFeePerGas || gasPrice;
+    assert(price, "Unable to get gas price");
+    return gasCost(this.defaultGas, price);
+  }
+  /**
+   * getTokenPrice. This should return the price of a token relative to the native gas token on the
+   * destination chain. This is tricky because the token address provided should be the token address
+   * on the destination chain, which may not be easily accessible when sending relay.
+   *
+   * @param {string} tokenAddress
+   * @returns {Promise<number | string>}
+   */
+  async getTokenPrice(tokenAddress: string): Promise<number | string> {
+    const [, tokenPrice] = await this.coingecko.getCurrentPriceByContract(tokenAddress, this.priceSymbol.toLowerCase());
+    return tokenPrice;
+  }
+  /**
+   * getTokenDecimals. Gets token decimals, its assumed decimals are the same on both chains.
+   *
+   * @param {string} tokenAddress
+   * @returns {Promise<number>}
+   */
+  async getTokenDecimals(tokenAddress: string): Promise<number> {
+    const erc20Client = erc20Connect(tokenAddress, this.provider);
+    return erc20Client.decimals();
+  }
+}
+
+interface RelayFeeCalculatorConfig {
+  discountPercent?: number;
+  feeLimitPercent?: number;
+  queries: QueryInterface;
+}
+
+export class RelayFeeCalculator {
+  private queries: Required<RelayFeeCalculatorConfig>["queries"];
+  private discountPercent: Required<RelayFeeCalculatorConfig>["discountPercent"];
+  private feeLimitPercent: Required<RelayFeeCalculatorConfig>["feeLimitPercent"];
+  constructor(config: RelayFeeCalculatorConfig) {
+    this.queries = config.queries;
+    this.discountPercent = config.discountPercent || 0;
+    this.feeLimitPercent = config.feeLimitPercent || 0;
+    assert(
+      this.discountPercent >= 0 && this.discountPercent <= 100,
+      "discountPercent must be between 0 and 100 percent"
+    );
+    assert(
+      this.feeLimitPercent >= 0 && this.feeLimitPercent <= 100,
+      "feeLimitPercent must be between 0 and 100 percent"
+    );
+  }
+  async relayerFeePercent(amountToRelay: BigNumberish, tokenAddress?: string): Promise<BigNumberish> {
+    const gasCosts = await this.queries.getGasCosts(tokenAddress);
+    // if not provided, its assumed the token is the same denomination as the destination gas token
+    if (!tokenAddress) {
+      return percent(gasCosts, amountToRelay).toString();
+    }
+    const tokenPrice = await this.queries.getTokenPrice(tokenAddress);
+    const decimals = await this.queries.getTokenDecimals(tokenAddress);
+    const gasFeesInToken = nativeToToken(gasCosts, tokenPrice, decimals);
```

Comment:
How would this work for matic? Who is telling nativeToToken not to convert to ETH, but to convert to Matic?

Reply:
each chain will have to implement their call to `getTokenPrice`, this price should convert whatever the native gas token is on the receiving chain to the token they are transferring. In most cases this will be eth/token, but in matics case you can change it to matic/token

Reply:
Okay, I think I understand where you're coming from. Makes sense to me now.